src/maingoblint.ml

```diff
@@ -318,6 +303,8 @@ let merge_preprocessed cpp_file_names =
   Cilfacade.current_file := merged_AST;
   merged_AST
 
+let preprocess_and_merge () = preprocess_files () |> merge_preprocessed
```

Comment:
Another issue that Karoliine stumbled upon is related to `merge_preprocessed` opening and setting `Goblintutil.out`, but that is never closed. This causes the `json-messages` output, which prints to that `out_channel` to append to the initially opened file, instead of overwriting its contents, yielding invalid JSON.

In the server mode, `Goblintutil.out` should at some point be closed as well (in non-server mode we just leave it up to the OS as the program exits). Then the subsequent opening would actually reopen the file and overwrite, instead of further appending to the already-opened file.

Reply:
Actually no, I got it backwards: reopening the same file each time in `merge_preprocessed` is not problematic, because the reopened descriptor truncates the file still. Although never closing it a small file descriptor leak.

The problem actually happens with `--disable server.reparse` because then `merge_preprocessed` is not called for reanalysis, so the same `Goblintutil.out` channel is opened and appended to. Opening of those channels should probably be pulled out of `merge_preprocessed` somewhere that gets re-run regardless of reparsing.

Reply:
Just wondering, but is an incremental run guaranteed/supposed to produce the same set of warnings as a full run? I moved the line that sets `Goblintutil.out` to `do_analyze` and now clear the message table before each run because otherwise it ends up containing stale warnings.

Reply:
Yes, the incremental analysis makes sure to output all the warnings again, so they go into the table. On `master`, this happens trivially through the postsolving/verify phase in the solver; on `interactive`, there's some special incremental warning logic that uses separate solver data structures to reuse some warnings. But either way they should all go back into the table, so clearing it is the way to go. Especially since in non-server incremental runs the messages table would be empty as well.